### PR TITLE
Export html with facteted curves

### PR DESCRIPTION
**Pull request by Shimwell**
## Proposed changes

This is progress towards getting the 2d image and 2d plotly graphs working.

They are still not quite perfect as they make use of the shape.face so I think they will remain a develop tool for checking the shapes and points look correct.

The reason they are not quite suitable for users is that the method does not account for boolean opperations prior to boolean operations.

However the main improvement is that the edges are now faceted for circles and splines with a controllable tolerance. This solves issue #84 

```python
import paramak

test_shape = paramak.RotateMixedShape(
    points=[
        (50, 0, "straight"),
        (50, 50, "spline"),
        (60, 70, "spline"),
        (70, 50, "circle"),
        (60, 25, "circle"),
        (70, 0, "straight")
        ],
    rotation_angle=10
)

test_shape.export_html()
test_shape.export_2d_image()
```

## before

Screenshot from 2021-01-25 00-13-23

## after

Screenshot from 2021-01-25 00-16-44

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] Pep8 applied
- [ ] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by RemDelaporteMathurin** [comment]
Great stuff @Shimwell ! 

About the Boolean operations. Is there a way to get the face wire from the solid attribute ?

Something like 

my_shape.solid.get_wire() 

?

File: paramak/reactor.py [hunk 1]
```diff
@@ -596,14 +597,21 @@ def export_2d_image(
...
 
         Args:
-            filename (str): the filename to save the html graph
+            filename: the filename used to save the html graph.
+            facet_splines: If True then spline edges will be faceted. Defaults
+                to True.
+            facet_splines: If True then circle edges will be faceted.Defaults
+                to True.
+            tolerance: faceting toleranceto use when faceting cirles and
+                splines. Defaults to 1e-3.
```
Comment (RemDelaporteMathurin) [inline paramak/reactor.py:611]: I don't see these args in the function arguments: is it a copy pasted code from shape.py ?
+ the types are missing

File: paramak/reactor.py [hunk 2]
```diff
@@ -620,7 +628,13 @@ def export_html(self, filename="reactor.html"):
 
         # accesses the Shape traces for each Shape and adds them to the figure
         for entry in self.shapes_and_components:
-            fig.add_trace(entry._trace())
+            edges = facet_wire(wire=entry.wire)
+            fpoints = []
+            for edge in edges:
+                for v in edge.Vertices():
```
Comment (RemDelaporteMathurin) [inline paramak/reactor.py:634]:
```suggestion
                for vertice in edge.Vertices():
```

File: paramak/reactor.py [hunk 3]
```diff
@@ -620,7 +628,13 @@ def export_html(self, filename="reactor.html"):
...
+                    fpoints.append((v.X, v.Z))
```
Comment (RemDelaporteMathurin) [inline paramak/reactor.py:635]:
```suggestion
                    fpoints.append((vertice.X, vertice.Z))
```

File: paramak/shape.py [hunk 4]
```diff
@@ -855,14 +856,27 @@ def export_svg(self, filename: str) -> str:
...
         Args:
-            filename (str): the filename used to save the html graph
+            filename: the filename used to save the html graph. Defaults to
+                shape.html
+            facet_splines: If True then spline edges will be faceted. Defaults
+                to True.
+            facet_splines: If True then circle edges will be faceted.Defaults
+                to True.
+            tolerance: faceting toleranceto use when faceting cirles and
+                splines. Defaults to 1e-3.
```
Comment (RemDelaporteMathurin) [inline paramak/shape.py:879]: - the types are missing
- `facet_splines `is written twice, second one needs to be replaced by `facet_circles` I suppose ?

File: paramak/shape.py [hunk 5]
```diff
@@ -1014,12 +1035,18 @@ def _create_patch(self):
...
-        for point in self.points:
-            xylist.append([point[0], point[1]])
+        edges = facet_wire(
+            wire=self.wire,
+            facet_splines=True,
+            facet_circles=True)
+
+        fpoints = []
+        for edge in edges:
+            for v in edge.Vertices():
```
Comment (RemDelaporteMathurin) [inline paramak/shape.py:1046]:
```suggestion
            for vertice in edge.Vertices():
```

File: paramak/shape.py [hunk 6]
```diff
@@ -1014,12 +1035,18 @@ def _create_patch(self):
...
+                fpoints.append((v.X, v.Z))
```
Comment (RemDelaporteMathurin) [inline paramak/shape.py:1047]:
```suggestion
                fpoints.append((vertice.X, vertice.Z))
```

**Comment by Shimwell** [comment]
> Great stuff @Shimwell !
> 
> About the Boolean operations. Is there a way to get the face wire from the solid attribute ?
> 
> Something like
> 
> my_shape.solid.get_wire()
> 
> ?

Yep we can get wires from the solid but the trouble is that we get all the wires. The current wire points to the wire made from the user input coordinates.

File: paramak/reactor.py (hunk 1 above)
Comment (Shimwell) [reply paramak/reactor.py]: Is it ok to have tpyes in the type hinting instead of in the docs

Comment (RemDelaporteMathurin) [reply paramak/reactor.py]: I think as long as users don't necessarly have to go to read the docs to know the types it's ok

File: paramak/shape.py [hunk 7]
```diff
@@ -885,21 +899,44 @@ def export_html(self, filename: str):
...
+            mode: str = "markers+lines"
+    ):
         """Creates a plotly trace representation of the points of the Shape
         object. This method is intended for internal use by Shape.export_html.
 
+        Args:
+            points: A list of tuples containing the X, Z points of to add to
+                the trace.
+            mode: The mode to use for the Plotly.Scatter graph. Options include
+                "markers", "lines" and "markers+lines". Defualts to
```
Comment (Shimwell) [inline paramak/shape.py:937]:
```suggestion
                "markers", "lines" and "markers+lines". Defaults to
```

**Comment by Shimwell** [comment]
Bit of progress on this PR. The reactors can now be exported as HTML. A few shapes that have their own create_solid methods needed the self.wire adding. There is a bit more tidying to do but good to see the tests are passing

**Comment by Shimwell** [comment]
I think the hollowcube (graveyard) does not work with this method as it doesn't have a .face or .points

**Comment by Shimwell** [comment]
The export_html can now handle different workplanes when exporting a shape and will draw the html in the workplane that the shape is made. The reactor export is still only available in the XZ plane

**Pull request merged**